File: rust/src/sieve.rs
pub fn sieve(size: usize) -> Vec<usize> {
    if size < 2 {
        return vec![];
    }

    let size2 = (size >> 1) as usize;
    let mut a = Vec::with_capacity(size2);
    a.resize(size2, 0);
    let q = (size as f64).sqrt().floor() as usize;
    let mut factor: usize = 3;

    while factor < q {
        for i in (factor..size).step_by(2) {
            if a[i >> 1] == 0 {
                factor = i;
                break;
            }
        }

        for i in (factor * factor..size).step_by(factor * 2) {
            a[i >> 1] = 1;
        }

        factor += 2;
    }

    a[0] = 2;
    let mut n_primes_found = 1;
    for i in (3..size).step_by(2) {
        if a[i >> 1] == 0 {
            a[n_primes_found] = i;
            n_primes_found += 1;
        }
    }
    a.resize(n_primes_found, 0);

    return a;
}

File: cpp/sieve.cpp
#include <cmath>

#include "sieve.h"

std::vector<unsigned int> sieve(unsigned int size) {
  if (size < 2)
    return {};

  // because all prime numbers (except 2) must be odd, we can bit-shift all
  // numbers to the right by 1 and reduce our memory usage by half
  const auto size_2 = size >> 1;
  auto a = std::vector<unsigned int>(size_2);
  const unsigned int q = std::sqrt(size);

  for (unsigned int factor = 3; factor < q; factor += 2) {
    for (unsigned int i = factor; i < size; i += 2) {
      if (!a[i >> 1]) {
        factor = i;
        break;
      }
    }

    for (unsigned int i = factor * factor; i < size; i += factor * 2) {
      a[i >> 1] = 1;
    }
  }

  a[0] = 2;
  unsigned int n_primes_found = 1;
  for (unsigned int i = 3; i < size; i += 2)
    if (!a[i >> 1])
      a[n_primes_found++] = i;

  a.erase(a.begin() + n_primes_found, a.end());
  return a;
}
